File: typescript/src/service.ts
import { createHmac } from "node:crypto";
import { WhatsAppClient } from "./client";
import { MessageHandler, WebhookHandler } from "./handlers";
import type {
  SendReactionParams,
  SendReactionResult,
  WhatsAppConfig,
  WhatsAppEventType,
  WhatsAppMessage,
  WhatsAppMessageResponse,
  WhatsAppWebhookEvent,
} from "./types";

type EventHandler = (payload: unknown) => void | Promise<void>;

/**
 * WhatsApp service for ElizaOS.
 * Provides comprehensive WhatsApp Cloud API integration.
 */
export class WhatsAppService {
  static readonly serviceType = "whatsapp";

  private config: WhatsAppConfig;
  private client: WhatsAppClient;
  private messageHandler: MessageHandler;
  private webhookHandler: WebhookHandler;
  private running = false;
  private eventHandlers: Map<WhatsAppEventType, EventHandler[]> = new Map();

  constructor(config: WhatsAppConfig) {
    this.config = config;
    this.client = new WhatsAppClient(config);
    this.messageHandler = new MessageHandler(this.client);
    this.webhookHandler = new WebhookHandler();
  }

  /**
   * Create service from environment variables.
   */
  static fromEnv(): WhatsAppService {
    const accessToken = process.env.WHATSAPP_ACCESS_TOKEN;
    const phoneNumberId = process.env.WHATSAPP_PHONE_NUMBER_ID;

    if (!accessToken) {
      throw new Error("WHATSAPP_ACCESS_TOKEN environment variable is required");
    }
    if (!phoneNumberId) {
      throw new Error("WHATSAPP_PHONE_NUMBER_ID environment variable is required");
    }

    return new WhatsAppService({
      accessToken,
      phoneNumberId,
      webhookVerifyToken: process.env.WHATSAPP_WEBHOOK_VERIFY_TOKEN,
      businessAccountId: process.env.WHATSAPP_BUSINESS_ACCOUNT_ID,
      apiVersion: process.env.WHATSAPP_API_VERSION || "v18.0",
    });
  }

  /**
   * Get the underlying client.
   */
  getClient(): WhatsAppClient {
    return this.client;
  }

  /**
   * Start the WhatsApp service.
   */
  async start(): Promise<void> {
    if (this.running) {
      console.warn("WhatsApp service is already running");
      return;
    }

    this.running = true;
    console.log(`WhatsApp service started for phone number ID: ${this.config.phoneNumberId}`);
  }

  /**
   * Stop the WhatsApp service.
   */
  async stop(): Promise<void> {
    if (!this.running) {
      return;
    }

    this.running = false;
    console.log("WhatsApp service stopped");
  }

  /**
   * Check if the service is running.
   */
  isRunning(): boolean {
    return this.running;
  }

  /**
   * Register an event handler.
   */
  onEvent(eventType: WhatsAppEventType, handler: EventHandler): void {
    const handlers = this.eventHandlers.get(eventType) || [];
    handlers.push(handler);
    this.eventHandlers.set(eventType, handlers);
  }

  /**
   * Handle a webhook event.
   */
  async handleWebhook(event: WhatsAppWebhookEvent): Promise<void> {
    await this.webhookHandler.handle(event);
  }

  /**
   * Verify a webhook subscription request.
   */
  verifyWebhook(mode: string, token: string, challenge: string): string | null {
    if (mode === "subscribe" && token === this.config.webhookVerifyToken) {
      console.log("Webhook verified successfully");
      return challenge;
    }
    console.warn("Webhook verification failed");
    return null;
  }

  /**
   * Verify webhook payload signature.
   */
  verifyWebhookSignature(payload: Buffer | string, signature: string): boolean {
    if (!this.config.webhookVerifyToken) {
      return true; // No verification configured
    }

    if (!signature.startsWith("sha256=")) {
      return false;
    }

    const expectedSignature = signature.slice(7);
    const computed = createHmac("sha256", this.config.webhookVerifyToken)
      .update(payload)
      .digest("hex");

    return computed === expectedSignature;
  }

  // Message sending methods

  /**
   * Send a message of any type.
   */
  async sendMessage(message: WhatsAppMessage): Promise<WhatsAppMessageResponse> {
    return this.messageHandler.send(message);
  }

  /**
   * Send a text message.
   */
  async sendText(to: string, text: string): Promise<WhatsAppMessageResponse> {
    const response = await this.client.sendTextMessage(to, text);
    return response.data;
  }

  /**
   * Send a reaction to a message.
   */
  async sendReaction(params: SendReactionParams): Promise<SendReactionResult> {
    return this.client.sendReaction(params);
  }

  /**
   * Remove a reaction from a message.
   */
  async removeReaction(to: string, messageId: string): Promise<SendReactionResult> {
    return this.client.removeReaction(to, messageId);
  }

  /**
   * Send an image message.
   */
  async sendImage(
    to: string,
    imageUrl: string,
    caption?: string
  ): Promise<WhatsAppMessageResponse> {
    const response = await this.client.sendImage(to, imageUrl, caption);
    return response.data;
  }

  /**
   * Send a video message.
   */
  async sendVideo(
    to: string,
    videoUrl: string,
    caption?: string
  ): Promise<WhatsAppMessageResponse> {
    const response = await this.client.sendVideo(to, videoUrl, caption);
    return response.data;
  }

  /**
   * Send an audio message.
   */
  async sendAudio(to: string, audioUrl: string): Promise<WhatsAppMessageResponse> {
    const response = await this.client.sendAudio(to, audioUrl);
    return response.data;
  }

  /**
   * Send a document message.
   */
  async sendDocument(
    to: string,
    documentUrl: string,
    filename?: string,
    caption?: string
  ): Promise<WhatsAppMessageResponse> {
    const response = await this.client.sendDocument(to, documentUrl, filename, caption);
    return response.data;
  }

  /**
   * Send a location message.
   */
  async sendLocation(
    to: string,
    latitude: number,
    longitude: number,
    name?: string,
    address?: string
  ): Promise<WhatsAppMessageResponse> {
    const response = await this.client.sendLocation(to, latitude, longitude, name, address);
    return response.data;
  }

  /**
   * Send an interactive button message.
   */
  async sendButtonMessage(
    to: string,
    bodyText: string,
    buttons: Array<{ id: string; title: string }>,
    headerText?: string,
    footerText?: string
  ): Promise<WhatsAppMessageResponse> {
    const response = await this.client.sendButtonMessage(
      to,
      bodyText,
      buttons,
      headerText,
      footerText
    );
    return response.data;
  }

  /**
   * Send an interactive list message.
   */
  async sendListMessage(
    to: string,
    bodyText: string,
    buttonText: string,
    sections: Array<{
      title?: string;
      rows: Array<{ id: string; title: string; description?: string }>;
    }>,
    headerText?: string,
    footerText?: string
  ): Promise<WhatsAppMessageResponse> {
    const response = await this.client.sendListMessage(
      to,
      bodyText,
      buttonText,
      sections,
      headerText,
      footerText
    );
    return response.data;
  }

  /**
   * Mark a message as read.
   */
  async markAsRead(messageId: string): Promise<boolean> {
    return this.client.markMessageAsRead(messageId);
  }

  /**
   * Get the download URL for a media file.
   */
  async getMediaUrl(mediaId: string): Promise<string | null> {
    return this.client.getMediaUrl(mediaId);
  }
}

File: rust/src/service.rs
//! WhatsApp service implementation

use crate::client::WhatsAppClient;
use crate::config::WhatsAppConfig;
use crate::error::{Result, WhatsAppError};
use crate::types::{IncomingMessage, WhatsAppChatState, WhatsAppMessageResponse, WhatsAppWebhookEvent};
use crate::WHATSAPP_SERVICE_NAME;
use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::RwLock;
use tracing::{debug, info, warn};

/// WhatsApp service for ElizaOS
pub struct WhatsAppService {
    client: Arc<RwLock<Option<WhatsAppClient>>>,
    config: Arc<RwLock<Option<WhatsAppConfig>>>,
    chat_states: Arc<RwLock<HashMap<String, WhatsAppChatState>>>,
    is_running: Arc<RwLock<bool>>,
}

impl WhatsAppService {
    /// Creates a new service instance
    pub fn new() -> Self {
        Self {
            client: Arc::new(RwLock::new(None)),
            config: Arc::new(RwLock::new(None)),
            chat_states: Arc::new(RwLock::new(HashMap::new())),
            is_running: Arc::new(RwLock::new(false)),
        }
    }

    /// Creates a service with the given configuration
    pub fn with_config(config: WhatsAppConfig) -> Result<Self> {
        config.validate()?;
        let client = WhatsAppClient::new(config.clone());
        
        Ok(Self {
            client: Arc::new(RwLock::new(Some(client))),
            config: Arc::new(RwLock::new(Some(config))),
            chat_states: Arc::new(RwLock::new(HashMap::new())),
            is_running: Arc::new(RwLock::new(false)),
        })
    }

    /// Gets the client reference
    pub async fn client(&self) -> Option<WhatsAppClient> {
        let config = self.config.read().await;
        config.as_ref().map(|c| WhatsAppClient::new(c.clone()))
    }

    /// Checks if the service is running
    pub async fn is_running(&self) -> bool {
        *self.is_running.read().await
    }

    /// Initializes the service
    pub async fn initialize(&self) -> Result<()> {
        let config = match WhatsAppConfig::from_env() {
            Ok(c) => c,
            Err(e) => {
                warn!("WhatsApp configuration not available: {}", e);
                return Err(e);
            }
        };

        if !config.enabled {
            info!("WhatsApp plugin is disabled via configuration");
            return Ok(());
        }

        let client = WhatsAppClient::new(config.clone());

        *self.config.write().await = Some(config);
        *self.client.write().await = Some(client);
        *self.is_running.write().await = true;

        info!("WhatsApp service started");
        Ok(())
    }

    /// Stops the service
    pub async fn stop(&self) {
        *self.is_running.write().await = false;
        *self.client.write().await = None;
        info!("WhatsApp service stopped");
    }

    /// Sends a text message
    pub async fn send_message(&self, to: &str, text: &str) -> Result<WhatsAppMessageResponse> {
        let client_guard = self.client.read().await;
        let client = client_guard
            .as_ref()
            .ok_or_else(|| WhatsAppError::config("Client not initialized"))?;

        client.send_text(to, text).await
    }

    /// Handles a webhook event
    pub async fn handle_webhook(&self, event: WhatsAppWebhookEvent) -> Result<()> {
        for entry in event.entry {
            for change in entry.changes {
                if change.field == "messages" {
                    if let Some(messages) = change.value.messages {
                        for message in messages {
                            self.handle_incoming_message(&message, &change.value.metadata.phone_number_id)
                                .await?;
                        }
                    }
                    
                    // Update contact info
                    if let Some(contacts) = change.value.contacts {
                        for contact in contacts {
                            let state = WhatsAppChatState {
                                phone_number_id: change.value.metadata.phone_number_id.clone(),
                                contact_wa_id: contact.wa_id.clone(),
                                contact_name: Some(contact.profile.name),
                                last_message_at: Some(chrono::Utc::now().timestamp()),
                            };
                            self.chat_states.write().await.insert(contact.wa_id, state);
                        }
                    }
                }
            }
        }
        Ok(())
    }

    async fn handle_incoming_message(&self, message: &IncomingMessage, phone_number_id: &str) -> Result<()> {
        info!(
            "Received message from {} (type: {})",
            message.from, message.message_type
        );

        // Extract text content
        let text = message.text.as_ref().map(|t| t.body.as_str());

        if let Some(text) = text {
            debug!("Message text: {}", text);
        }

        // Update chat state
        let state = WhatsAppChatState {
            phone_number_id: phone_number_id.to_string(),
            contact_wa_id: message.from.clone(),
            contact_name: None,
            last_message_at: Some(
                message
                    .timestamp
                    .parse::<i64>()
                    .unwrap_or_else(|_| chrono::Utc::now().timestamp()),
            ),
        };
        self.chat_states.write().await.insert(message.from.clone(), state);

        Ok(())
    }

    /// Verifies a webhook token
    pub async fn verify_webhook(&self, token: &str) -> bool {
        let client_guard = self.client.read().await;
        client_guard
            .as_ref()
            .map(|c| c.verify_webhook(token))
            .unwrap_or(false)
    }

    /// Gets chat state for a contact
    pub async fn get_chat_state(&self, wa_id: &str) -> Option<WhatsAppChatState> {
        self.chat_states.read().await.get(wa_id).cloned()
    }
}

impl Default for WhatsAppService {
    fn default() -> Self {
        Self::new()
    }
}

impl WhatsAppService {
    /// Get the service name
    pub fn name(&self) -> &str {
        WHATSAPP_SERVICE_NAME
    }

    /// Get the service description
    pub fn description(&self) -> &str {
        "WhatsApp Cloud API service"
    }
}
